pyproject.toml

```diff
@@ -0,0 +1,6 @@
+[build-system]
+requires = [
+	"setuptools~=62.0.0",
+	"wheel~=0.37.0",
```

Comment:
What's the reasoning behind these versions?

Also, I assume this file is now read during the call to `python -m build` ?

Reply:
> What's the reasoning behind these versions?

These were the latest versions when I first wrote the `pyproject.toml` in #499. The fact that they are now out of date (`setuptools` is on `65.5.1`) highlights the slight maintainability burden here. I feel like it is safer to pin (maybe to minor instead?) but can't see a way to avoid these going out of date then...

> Also, I assume this file is now read during the call to python -m build ?

Yep, the idea here is that `build` (and `pip install`) reads these deps, and then installs them in an isolated virtual env, hence giving us more control over the build env. Since we are now building for release via GA, we could of course define the build env via a `requirements.txt`. However, the `pyproject.toml` approach gives more consistency between the build environment for PyPI and that used when users install locally.

Reply:
Sidenote: The caveat of this approach is that `build-system` defines the backend. There isn't currently a way to define the frontend deps (i.e. the `build` version) in the `pyproject.toml`, hence why I am having to define the version for that in the `Makefile` and GA yaml.

Reply:
I see what you mean wrt dependencies going out of date if pinned, but for release builds it shouldn't matter (and also note that the frontend `build` in the GA is now pinned exactly). On the other hand, the minimum version via `~` makes sense to me too, though that could have a small risk of breaking things. TBF I think leaving `~` here is fine.

Btw when you say the `pyproject.toml` deps are installed in an isolated virtual enviroment, how does that work exactly? Assuming I'm already in a virtualenv/conda env, would in this case, after calling `pip install .`, `setuptools` and `wheel` be installed in a new (not visible to user?) virtualenv,  build takes place in that env, and then the package is installed in the original user env?

Reply:
> `setuptools` and `wheel` be installed in a new (not visible to user?) virtualenv, build takes place in that env, and then the package is installed in the original user env?

Yes, from what I understand, exactly this. For packages conforming to [PEP 518](https://peps.python.org/pep-0518/) (i.e. ones that spec their `build-system` in a `pyproject.toml`, running `pip install` creates a virtual env hidden from the user, installs build deps, does the build, and then installs the built package in the users own environment (i.e. their system python, or conda/virtualenv etc). When running `pip install .` with this PR we get:

```
Processing /home/ascillitoe/Software/alibi-detect
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
```

Note the extra lines compared to when we run the same with our v0.10.4:

```
Processing /home/ascillitoe/Software/alibi-detect
  Preparing metadata (setup.py) ... done
```

**Note**: You can disable the above and use your already-installed build tools with the `pip` [--no-build-isolation](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-no-build-isolation) flag. Described as:

> Disable isolation when building a modern source distribution. Build dependencies specified by PEP 518 must be already installed if this option is used.